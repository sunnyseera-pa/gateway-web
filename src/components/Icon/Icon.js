/** @jsx jsx */
import { cx } from '@emotion/css';
import { jsx } from '@emotion/react';
import PropTypes from 'prop-types';
import { addCommonPropTypes } from '../../configs/propTypes';
import useCommonStyles from '../../hooks/useCommonStyles';
import * as styles from './Icon.styles.js';

const Icon = ({ svg, size, color, fill, stroke, className, ml, mr, mb, mt, ...outerProps }) => {
    const commonStyles = useCommonStyles({ mt, mb, ml, mr });

    const iconProps = {
        css: styles.root({ size, color, fill, stroke }),
        className: cx('ui-Icon', className, commonStyles),
        ...outerProps,
    };

    return <span {...iconProps}>{svg}</span>;
};

Icon.propTypes = addCommonPropTypes({
    svg: PropTypes.node.isRequired,
    size: PropTypes.oneOf(['xxs', 'xs', 'sm', 'md', 'default', 'lg', 'xl', '2xl', '3xl', '4xl', '5xl', 'contained']),
    color: PropTypes.string,
    fill: PropTypes.string,
    stroke: PropTypes.string,
    inline: PropTypes.bool,
});

Icon.defaultProps = {
    size: 'default',
    color: 'inherit',
    fill: 'inherit',
    stroke: 'none',
};

export default Icon;
