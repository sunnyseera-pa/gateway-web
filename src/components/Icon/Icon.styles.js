import { css } from '@emotion/react';

export const root = ({ size, color, stroke, fill }) => theme => {
	const {
		colors,
		components: {
			Icon: { sizes },
		},
	} = theme;

	return css`
		display: inline-flex;
		stroke: ${stroke};
		color: ${colors[color]};
		fill: ${colors[fill]};
		height: ${sizes[size]};
		width: ${sizes[size]};
		justify-content: center;
		align-items: center;

		svg,
		img {
			width: 100%;
			height: 100%;
		}
	`;
};
