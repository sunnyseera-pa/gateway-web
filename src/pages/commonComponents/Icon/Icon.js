/** @jsx jsx */
import { cx } from '@emotion/css';
import { jsx } from '@emotion/react';
import axios from 'axios';
import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import { addCommonPropTypes } from '../../../configs/propTypes';
import useCommonStyles from '../../../hooks/useCommonStyles';
import SVGIcon from '../../../images/SVGIcon';
import * as styles from './Icon.styles.js';

const svgFragments = {
	dataseticon: '0 0 24 24',
	projecticon: '0 0 24 24',
	searchicon: '0 0 18 18',
	cohorticon: '1 1 10 10',
	newnotificationicon: '9 5 6 6',
	toolicon: '0 0 18 18',
	emptystar: '0 0 24 24',
	clear: '0 0 24 24',
	stock: '0 0 24 24',
	newprojecticon: '0 0 24 24',
	newtoolicon: '0 0 24 24',
	reviewsicon: '0 0 24 24',
	accounticon: '0 0 24 24',
	dataaccessicon: '0 0 24 24',
	rolesicon: '0 0 24 24',
	bell: '0 0 24 24',
	newestprojecticon: '0 0 24 24',
	personicon: '2 4 20 20',
	closeicon: '3 3 18 18',
	collections: '0 0 24 24',
	chevronbottom: '4 4 17 17',
	chevronright: '3 3 18 18',
	dashboard: '0 0 24 24',
	info: '0 0 24 24',
	infofill: '0 0 24 24',
	chat: '0 0 24 24',
	check: '0 0 24 24',
	workflow: '0 0 24 24',
	attention: '0 0 16 16',
	eye: '0 0 24 24',
	eyeCrossed: '0 0 24 24',
	educationicon: '0 -1 26 26',
	members: '6 8 16 16',
	shield: '3 3 18 18',
	externallink: '0 0 24 24',
	checkicon: '0 0 24 24',
	loading: '0 0 24 24',
	cycle: '0 0 24 24',
	flag: '0 0 24 24',
	'metadata-complete': '0 0 16 16',
	'metadata-half-done': '0 0 16 16',
	'metadata-empty': '0 0 16 16',
	plusChunky: '6 6 24 24',
	tick: '6 6 12 12',
};

const Icon = ({ name, size, color, fill, stroke, className, ml, mr, mb, mt, inline, ...outerProps }) => {
	const [svg, setSvg] = useState();
	const commonStyles = useCommonStyles({ mt, mb, ml, mr });

	useEffect(() => {
		const importIcon = () => {
			import(`../../../images/${name}.svg`).then(({ default: namedImport }) => {
				axios.get(namedImport).then(({ data }) => {
					setSvg(data);
				});
			});
		};

		if (name) importIcon();
	}, [name]);

	const iconProps = {
		css: styles.root({ size, color, fill, stroke }),
		className: cx('ui-Icon', className, commonStyles),
		...outerProps,
	};

	if (inline) {
		return (
			<span {...iconProps}>
				<SVGIcon name={name} fill='inherit' color='inherit' stroke='inherit' viewBox={svgFragments[name]} />
			</span>
		);
	} else if (!inline && svg) {
		return <span {...iconProps} dangerouslySetInnerHTML={{ __html: svg }} />;
	}

	return null;
};

Icon.propTypes = addCommonPropTypes({
	name: PropTypes.string.isRequired,
	size: PropTypes.oneOf(['xxs', 'xs', 'sm', 'md', 'default', 'lg', 'xl', '2xl', '3xl', '4xl', '5xl']),
	color: PropTypes.string,
	fill: PropTypes.string,
	stroke: PropTypes.string,
	inline: PropTypes.bool,
});

Icon.defaultProps = {
	size: 'default',
	color: 'inherit',
	fill: 'inherit',
	stroke: 'none',
	inline: false,
};

export default Icon;
