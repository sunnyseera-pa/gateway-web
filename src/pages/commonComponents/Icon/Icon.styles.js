export const root =
	({ size, color, fill, ...commonProps }) =>
	({
		colors,
		components: {
			Icon: { sizes },
		},
	}) =>
		addCommonProps(
			commonProps,
			css`
				color: ${colors[color]};
				fill: ${colors[fill]};
				height: ${sizes[size]};
				width: ${sizes[size]};
			`
		);
