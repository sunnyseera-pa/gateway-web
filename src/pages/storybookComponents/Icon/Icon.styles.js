import { css } from '@emotion/react';

export const root = ({ size, color, stroke, fill }) => theme => {
	const {
		colors,
		components: {
			Icon: { sizes },
		},
	} = theme;

	return css`
		display: inline-flex;
		stroke: ${stroke};
		color: ${colors[color]};
		fill: ${colors[fill]};
		height: ${sizes[size]};
		justify-content: center;
		align-items: center;

		svg,
		img {
			width: 100%;
			height: 100%;
		}
	`;
};
