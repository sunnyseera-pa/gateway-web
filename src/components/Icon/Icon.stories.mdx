import { Meta, Story, Canvas } from '@storybook/addon-docs';
import SVG from 'react-inlinesvg';
import starIcon from '../../images/star.svg';

import Icon from './Icon';

<Meta
    title='Core/Icon'
    component={Icon}
    args={{
        ...Icon.defaultProps,
    }}
    argTypes={{
        color: {
            control: 'text',
        },
    }}
/>

<Canvas>
    <Story name='Sizes'>
        {props => {
            return ['xxs', 'xs', 'sm', 'md', 'lg', 'xl', '2xl', '3xl', '4xl', '5xl'].map(size => (
                <Icon svg={<SVG src={starIcon} />} {...props} size={size} mr={2} />
            ));
        }}
    </Story>
</Canvas>
