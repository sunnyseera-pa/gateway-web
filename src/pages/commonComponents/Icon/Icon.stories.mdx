import { Meta, Story, Canvas } from '@storybook/addon-docs';

import Icon from './Icon';

<Meta
	title='Icon'
	component={Icon}
	args={{
		name: 'tick',
	}}
/>

<Canvas>
	<Story name='Basic'>{props => <Icon {...props} />}</Story>
</Canvas>
