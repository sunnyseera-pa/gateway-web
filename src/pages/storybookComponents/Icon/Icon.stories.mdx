import { Meta, Story, Canvas } from '@storybook/addon-docs';

import Icon from './Icon';

<Meta
	title='Icon'
	component={Icon}
	args={{
		name: 'tick',
		...Icon.defaultProps,
	}}
	argTypes={{
		size: {
			options: ['xxs', 'xs', 'sm', 'md', 'default', 'lg', 'xl', '2xl', '3xl', '4xl', '5xl'],
			control: { type: 'select' },
		},
		color: {
			control: 'text',
		},
	}}
/>

<Canvas>
	<Story name='Basic'>{props => <Icon {...props} />}</Story>
</Canvas>
