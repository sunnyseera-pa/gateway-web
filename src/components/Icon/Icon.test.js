import React from 'react';
import { render, waitFor } from '@testing-library/react';
import axios from 'axios';
import Icon from '.';

jest.mock('axios');

jest.mock('../../images/SVGIcon', () => props => {
    mockSVGIcon(props);
    return <div />;
});

const mockSVGIcon = jest.fn();

let wrapper;

describe('Given the Icon component', () => {
    describe('When it is from a file', () => {
        beforeAll(async () => {
            axios.get.mockImplementation(() => Promise.resolve({ data: '<svg />' }));

            wrapper = render(<Icon svg='Application_approved' />, {
                wrapper: Providers,
            });

            await waitFor(() => expect(wrapper.container.querySelector('svg')).toBeTruthy());
        });

        it('Then matches the previous snapshot', () => {
            expect(wrapper.container).toMatchSnapshot();
        });
    });
});
