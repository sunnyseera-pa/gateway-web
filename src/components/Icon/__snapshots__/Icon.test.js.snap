// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the Icon component When it is from a file Then matches the previous snapshot 1`] = `
<div>
  <span
    class="ui-Icon css-1sivg29 css-1nyc6qc-Icon"
  >
    Application_approved
  </span>
</div>
`;
